src/txn/txn_rollback_to_stable.c

```diff
@@ -346,17 +346,17 @@ __rollback_ondisk_fixup_key(WT_SESSION_IMPL *session, WT_REF *ref, WT_PAGE *page
 
         /*
          * Do not include history store updates greater than on-disk data store version to construct
-         * a full update to restore. Include the most recent updates than the on-disk version
-         * shouldn't be problem as the on-disk version in history store is always a full update. It
-         * is better to not to include those updates as it unnecessarily increases the rollback to
-         * stable time.
+         * a full update to restore except when the on-disk update is prepared. Include the most
+         * recent updates than the on-disk version shouldn't be problem as the on-disk version in
+         * history store is always a full update. It is better to not to include those updates as it
+         * unnecessarily increases the rollback to stable time.
          *
          * Comparing with timestamps here has no problem unlike in search flow where the timestamps
          * may be reset during reconciliation. RTS detects an on-disk update is unstable based on
          * the written proper timestamp, so comparing against it with history store shouldn't have
          * any problem.
          */
-        if (hs_start_ts <= unpack->tw.start_ts) {
+        if (hs_start_ts <= unpack->tw.start_ts || unpack->tw.prepare) {
             if (type == WT_UPDATE_MODIFY)
```

Comment:
May be we can assert it is not prepared here.

Reply:
If we assert it can go wrong. As the full_value is first taken from the on-disk. it must be replaced with the proper value in the history store. If we don't go in that code flow, we restore an update with the timestamp of the history store and the on-disk data.

Reply:
I mean if the type is modify, then the onpage value can't be prepared.

Reply:
Understood. I also thought of adding an assert, but it will hit if the restored update is a modify.
Like scenario, P(20) -> U3(50)->U2(40)->M(30). Stable timestamp is 30. In this scenario the Modify is restored.